massa-execution-worker/src/execution.rs

```diff
@@ -0,0 +1,501 @@
+
+        // note that here, async operations should be executed
+
+        // check if there is a block at this slot
+        if let (Some((block_id, block)), Some(block_creator_addr)) =
+            (opt_block, opt_block_creator_addr)
+        {
+            // Try executing the operations of this block in the order in which they appear in the block.
+            // Errors are logged but do not interrupt the execution of the slot.
+            for (op_idx, operation) in block.operations.iter().enumerate() {
+                if let Err(err) = self.execute_operation(operation, block_creator_addr) {
+                    debug!(
+                        "failed executing operation index {} in block {}: {}",
+                        op_idx, block_id, err
+                    );
+                }
+            }
+        }
+
+        // return the execution output
+        context_guard!(self).take_execution_output()
+    }
+
+    /// Execues a read-only execution request.
+    /// The executed bytecode appears to be able to read and write the consensus state,
+    /// but all accumulated changes are simply returned as an ExecutionOutput object,
+    /// and not actually applied to the consensus state.
+    ///
+    /// # Arguments
+    /// * req: a read-only execution request
+    ///
+    /// # Returns
+    ///  ExecutionOutput describing the output of the execution, or an error
+    pub(crate) fn execute_readonly_request(
+        &self,
+        req: ReadOnlyExecutionRequest,
+    ) -> Result<ExecutionOutput, ExecutionError> {
+        // set the execution slot to be the one after the latest executed active slot
+        let slot = self
+            .active_cursor
+            .get_next_slot(self.config.thread_count)
+            .expect("slot overflow in readonly execution");
+
+        // accumulate ledger changes that happened in the output history before this slot
+        let previous_ledger_changes = self.get_accumulated_active_changes_at_slot(slot);
+
+        // create a readonly execution context
+        let max_gas = req.max_gas;
+        let bytecode = req.bytecode.clone();
+        let execution_context = ExecutionContext::new_readonly(
+            slot,
+            req,
+            previous_ledger_changes,
+            self.final_ledger.clone(),
+        );
+
+        // set the execution context for execution
+        *context_guard!(self) = execution_context;
+
+        // run the intepreter
+        massa_sc_runtime::run(&bytecode, max_gas, &*self.execution_interface)
+            .map_err(|err| ExecutionError::RuntimeError(err.to_string()))?;
+
+        // return the execution output
+        Ok(context_guard!(self).take_execution_output())
+    }
+
+    /// Gets a full ledger entry both at the latest final and active executed slots
+    /// TODO: this can be heavily optimized, see comments
+    ///
+    /// # returns
+    /// (final_entry, active_entry)
+    pub fn get_full_ledger_entry(
```

Comment:
```suggestion
    pub fn get_final_and_active_ledger_entries(
```

But I'd rather break that into two functions